tools/dep-collector/gobuild.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	gb "go/build"
+	"path/filepath"
+	"strings"
+
+	"knative.dev/pkg/test/cmd"
+)
+
+// https://golang.org/pkg/cmd/go/internal/modinfo/#ModulePublic
+type modInfo struct {
+	Path string
+	Dir  string
+}
+
+type gobuild struct {
+	mod *modInfo
+}
+
+// moduleInfo returns the module path and module root directory for a project
+// using go modules, otherwise returns nil.
+// If there is something wrong in parsing the json output, it will return an error.
+//
+// Related: https://github.com/golang/go/issues/26504
+func moduleInfo() (*modInfo, error) {
+	output, err := cmd.RunCommand("go list -mod=readonly -m -json")
+	if err != nil {
+		return nil, nil
```

Comment:
what if go doesn't support `-m -json` one day? Is there a way to make this check more strict? i.e. check if there is a command line exception

Reply:
I don't think there is a reliable way to detect if an option is supported or not, as the error code returned by the command is tied to the implementation details.

The `-m` flag is used to list modules instead of packages and I believe there is minor chance for it to be changed. I can run `go list -m` to check if it's using Go modules or not, and then run the `go list -mod=readonly -m -json` to get the json output. If you think it's okay I'll make the change.

Reply:
This would be preferable to me. The underline issue of invoking command is like life time binding, and it breaks without us being prepared. I think it's ok to return nil and nil for `go list -m` or make it more explicit by wrapping it inside `function isMod`. Anything failed afterwards should be treated as errors

Reply:
Makes sense. Done.